### PR TITLE
bpo-40631: Disallow star targets with parentheses around

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue40631

### CODE REVIEW

**Comment by pablogsal** [comment]
I am not completely happy about this approach but is late here and I cannot see immediately a better way without refactoring the whole star_targets.

File: Grammar/python.gram [hunk 1]
```diff
@@ -578,6 +578,8 @@ kwarg_or_double_starred[KeywordOrStarred*]:
 # NOTE: star_targets may contain *bitwise_or, targets may not.
 star_targets[expr_ty]:
     | a=star_target !',' { a }
+    | '(' a=starred_expression ')' {
+        RAISE_SYNTAX_ERROR_KNOWN_LOCATION(a, "starred assignment target must be in a list or tuple") }
```
Comment (pablogsal) [inline Grammar/python.gram:582]: If we prefer to also have:

```
*a = 1,2,3
```

raise `starred assignment target cannot be used with parentheses around`, then we can remove this line.

**Review by lysnikolaou — COMMENTED** [review]
Let's not break the rule that error messages only get generated on the second pass. I already got a refactoring for `star_targets` in mind, so I could take it up.

**Comment by pablogsal** [comment]
> Let's not break the rule that error messages only get generated on the second pass. I already got a refactoring for `star_targets` in mind, so I could take it up.

The base of the problem here is that if we don't immediately fail here the parsing continues and some other rule in the targets succeed. In particular it allows to recurse into the rule that accept a star target surrounded by parentheses. In other words: if we don't explicitly fail there is another rule that will succeed, as opposed to the parsing failing.

**Comment by lysnikolaou** [comment]
> > Let's not break the rule that error messages only get generated on the second pass. I already got a refactoring for `star_targets` in mind, so I could take it up.
> 
> The base of the problem here is that if we don't immediately fail here the parsing continues and some other rule in the targets succeed. In particular it allows to recurse into the rule that accept a star target surrounded by parentheses. In other words: if we don't explicitly fail there is another rule that will succeed, as opposed to the parsing failing.

Let me quickly open a PR to show what I've got so far, which seems to be working.

**Comment by lysnikolaou** [comment]
Here it is: #24027.

**Comment by pablogsal** [comment]
> Here it is: #24027.

Nice! I like that approach better. it's up to you if you want to move the tests in my PR to yours or change the grammar file here to use your approach :)

**Comment by lysnikolaou** [comment]
Cool cool. I've moved all your tests to that PR. I'll let you close this.

**Pull request closed without merging**